File: Program's_Contributed_By_Contributors/Java_Programs/Sorting/HeapSort.java
// Heap Sort
public class HeapSort
{
    public void sort(int arr[])
    {
        int n = arr.length;

        for (int i = n / 2 - 1; i >= 0; i--)
            inheap(arr, n, i);

        for (int i=n-1; i>=0; i--)
        {

            int temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;


            inheap(arr, i, 0);
        }
    }


    void inheap(int arr[], int n, int i)
    {
        int largest = i;
        int l = 2*i + 1;
        int r = 2*i + 2;

        if (l < n && arr[l] > arr[largest])
            largest = l;

        if (r < n && arr[r] > arr[largest])
            largest = r;

        if (largest != i)
        {
            int swap = arr[i];
            arr[i] = arr[largest];
            arr[largest] = swap;


            inheap(arr, n, largest);
        }
    }

    static void printArray(int arr[])
    {
        int n = arr.length;
        for (int i=0; i<n; ++i)
            System.out.print(arr[i]+" ");
        System.out.println();
    }


    public static void main(String args[])
    {
        int arr[] = {12, 11, 13, 5, 6, 7};
        int n = arr.length;

        HeapSort ob = new HeapSort();
        ob.sort(arr);

        System.out.println("Sorted array is");
        printArray(arr);
    }
}

File: Program's_Contributed_By_Contributors/Python_Programs/heap sort.py
import heapq

def heapify(array, n, i):
  """Converts a subtree into a max heap.
  Args:
    array: The array to heapify.
    n: The size of the heap.
    i: The index of the root node of the subtree.
  """
  largest = i
  left = 2 * i + 1
  right = 2 * i + 2

  if left < n and array[left] > array[largest]:
    largest = left
  if right < n and array[right] > array[largest]:
    largest = right

  if largest != i:
    array[i], array[largest] = array[largest], array[i]
    heapify(array, n, largest)

def heapsort(array):
  """Sorts the array using heap sort.
  Args:
    array: The array to sort.
  """
  n = len(array)

  # Build a max heap.
  for i in range(n // 2 - 1, -1, -1):
    heapify(array, n, i)

  # Extract the elements from the heap one by one.
  for i in range(n - 1, 0, -1):
    array[0], array[i] = array[i], array[0]
    heapify(array, i, 0)

array = [5, 3, 2, 1, 4]
heapsort(array)
print("Sorted array:", array)
